### DISCUSSION TITLE
onMount cleanup method is called right after mount on React dev mode due to react 18

### DISCUSSION THREAD

**Discussion by omerman** [pmndrs/jotai#2180] [Bug report]
## Summary
the cleanup(unmount) method for onMount of atoms is called right after the atom is initiated in Dev mode.
This creates issues if you set data in the cleanup method, because it's not the intended behavior.

* I get that you generally might not want to set data in the cleanup, but if we take a usecase when we have atomFamily and we want to remove a record from there onunmount, we should be able to do it freely, right?

## Link to reproduction
https://codesandbox.io/s/zealous-kowalevski-9f6n22?file=/src/App.tsx

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

I wrote a solution that works for me
```
import { Atom } from 'jotai';


const devCleanups = new Set<Atom<unknown>>();
export const atomCleanup = (atom: Atom<unknown>, fn: () => void) => {
  return () => {
    // React 18 calls useEffect twice in dev mode, so this code protects us.
    if (process.env.NODE_ENV === 'development') {
      if (!devCleanups.has(atom)) {
        // In the first time only add to set.
        devCleanups.add(atom);
        return;
      } else {
        // second time means we truly are trying to clean up the atom.
        devCleanups.delete(atom);
      }
    }

    fn();
  };
};
```

Please let me know what you think

**Comment by dai-shi** [comment]
> but if we take a usecase when we have atomFamily and we want to remove a record from there onunmount

Even if you feel the use case is valid, the double effect behavior can happen in production in the future. That's why it behaves so in dev mode. If you don't plan to update React version, it's fine, but usually it's not the case, right?
